### PR TITLE
Support SIGINT in FreqAI

### PR DESCRIPTION
**Pull request by wizrds**
## Summary

Support shutting down FreqAI and child threads when received SIGINT. This will allow the current training iteration to finish, but will close after without allowing more to start.

### CODE REVIEW

File: freqtrade/strategy/interface.py [hunk 1]
```diff
@@ -168,6 +168,10 @@ def start(self, *args, **kwargs):
                     raise OperationalException(
                         'freqAI is not enabled. '
                         'Please enable it in your config to use this strategy.')
+
+                def cleanup(self, *args, **kwargs):
```
Comment (xmatthias) [inline freqtrade/strategy/interface.py]: the name of this method should be aligned to the method you call in `ft_bot_cleanup()` - which is shutdown - not `cleanup()`.

File: freqtrade/freqtradebot.py [hunk 2]
```diff
@@ -148,6 +148,8 @@ def cleanup(self) -> None:
 
         self.check_for_open_trades()
 
+        self.strategy.ft_bot_cleanup()
```
Comment (xmatthias) [inline freqtrade/freqtradebot.py:151]: should probably also be called from backtesting ...

**Review by xmatthias — COMMENTED** [review]
While i understand what you're trying to do - is this really an issue (it's not one i noticed so far)?

the tread is not started as daemon - so once the main thread exits, child threads should clean up, too.
from [the docs](https://docs.python.org/3/library/threading.html#threading.Thread.daemon): 
> the main thread is not a daemon thread and therefore all threads created in the main thread default to daemon = False.
> The entire Python program exits when no alive non-daemon threads are left.

File: freqtrade/strategy/interface.py (hunk 1 above)
Comment (wizrds) [reply freqtrade/strategy/interface.py]: Good catch, originally called it cleanup but changed it to shutdown. Forgot about that part.

**Comment by wizrds** [comment]
> While i understand what you're trying to do - is this really an issue (it's not one i noticed so far)?
> 
> the tread is not started as daemon - so once the main thread exits, child threads should clean up, too. from [the docs](https://docs.python.org/3/library/threading.html#threading.Thread.daemon):
> 
> > the main thread is not a daemon thread and therefore all threads created in the main thread default to daemon = False.
> > The entire Python program exits when no alive non-daemon threads are left.

So, yes when CTRL+C was pressed, the child threads would continue as it was an infinite loop. This meant that telegram's `/reload_config` would not work either. As well, @robcaulk had mentioned an elusive bug that coincided with this where people would try and shut down in the middle of pickling the model to disk, which would cause errors. I think key word there is `no alive`, as the `start_scanning` thread would still be alive even when the main thread signaled a close.

**Comment by xmatthias** [comment]
> As well, @robcaulk had mentioned an elusive bug that coincided with this where people would try and shut down in the middle of pickling the model to disk,

nothing will change in this regards though - i can still force-exit the process by pressing ctrl+c multiple times - unless that's explicitly delayed until pickling is finished - i could still corrupt the file.

File: freqtrade/freqtradebot.py (hunk 2 above)
Comment (wizrds) [reply freqtrade/freqtradebot.py:151]: Where would that be called? I can't seem to find where things are cleaned up in the `Backtesting` object.

**Comment by wizrds** [comment]
> > As well, @robcaulk had mentioned an elusive bug that coincided with this where people would try and shut down in the middle of pickling the model to disk,
> 
> nothing will change in this regards though - i can still force-exit the process by pressing ctrl+c multiple times - unless that's explicitly delayed until pickling is finished - i could still corrupt the file.

No it doesn't directly fix it completely, though it should help with some things. At the minimum, at least when SIGINT is sent the first time the user can expect it to shutdown soon, rather than never.

File: freqtrade/freqtradebot.py (hunk 2 above)
Comment (wizrds) [reply freqtrade/freqtradebot.py:151]: Just at the bottom of `Backtesting.start`?

Comment (robcaulk) [reply freqtrade/freqtradebot.py:151]: > should probably also be called from backtesting ...

Can you be more specific? FreqAI backtesting is all single threaded. It does not require this kind of shut down sequence.

**Comment by robcaulk** [comment]
> > As well, @robcaulk had mentioned an elusive bug that coincided with this where people would try and shut down in the middle of pickling the model to disk,
> 
> nothing will change in this regards though - i can still force-exit the process by pressing ctrl+c multiple times - unless that's explicitly delayed until pickling is finished - i could still corrupt the file.

It is true, technically. But enabling /reload is important in itself. Which is also a goal here. Also, this is just cleaner and more correct.

With regards to a "full fix" I will eventually add a  backup which will always keep one non-corrupted backup of the `historic_predictions.pkl` available in case of corruption.

File: freqtrade/freqtradebot.py (hunk 2 above)
Comment (xmatthias) [reply freqtrade/freqtradebot.py:151]: well my assumption was that `ft_bot_cleanup()` may be used to clean up other things too over time.

if we discover something else that'll have to be shutdown (with or without freqAI context) - will you remember to check if this is actually called everywhere? I know i probably won't ... 

Leave it for now - @robcaulk  is right, we'll not need this in backtesting for now.

**Comment by xmatthias** [comment]
> But enabling /reload is important in itself. Which is also a goal here. Also, this is just cleaner and more correct.

that's not something that's obvious from the desc - but yeah, that's for sure an important point.

**Review by xmatthias — APPROVED** [review]
LGTM :+1: 
(i'll not be merging this for now - as i'd like @robcaulk's approval, too (feel free to merge if you're fine with it).

**Comment by robcaulk** [comment]
Wow this works surprisingly well. I am going to add the `historical_predictions.backup.pkl` then merge it :-)

**Review by robcaulk — APPROVED** [review]
LGTM

**Pull request merged**